dotnet crashes on rhel8 arm64 while unwinding stack

The CI job we have that runs the master branch on RHEL8 arm64 is crashing for each test process.

The building of .NET works fine. That is still using the .NET 6.0 sdk, so this is probably a regression.

Interestingly, the CI job that runs on RHEL9 arm64 doesn't have these crashes. One notable difference with RHEL8 which has caused issues with .NET in the past is RHEL8 having a 64k page size rather than the usual 4k.

When running a simple program like:
```
// See https://aka.ms/new-console-template for more information
Console.WriteLine("Hello, World!");
throw new Exception();
```

`gdb /root/repos/runtime/artifacts/bin/testhost/net7.0-Linux-Debug-arm64/dotnet` gives the following stacktrace which shows a zero pointer being dereferenced.
```
Hello, World!

Thread 1 "dotnet" received signal SIGSEGV, Segmentation fault.
access_mem (as=<optimized out>, addr=0, val=0xffffffffb090, write=<optimized out>, arg=<optimized out>)
    at /root/repos/runtime/src/native/external/libunwind/src/aarch64/Ginit.c:337
337	      *val = *(unw_word_t *) addr;
(gdb) bt
#0  access_mem (as=<optimized out>, addr=0, val=0xffffffffb090, write=<optimized out>, arg=<optimized out>)
    at /root/repos/runtime/src/native/external/libunwind/src/aarch64/Ginit.c:337
#1  0x0000ffffbdfb4728 in is_plt_entry (c=c@entry=0xffffffffb0e0) at /root/repos/runtime/src/native/external/libunwind/src/aarch64/Gstep.c:43
#2  0x0000ffffbdfb4670 in _ULaarch64_step (cursor=0xffffffffb0e0) at /root/repos/runtime/src/native/external/libunwind/src/aarch64/Gstep.c:171
#3  0x0000ffffbdf7aeec in PAL_VirtualUnwind (context=0xffffffffbd80, contextPointers=0xffffffffbce0)
    at /root/repos/runtime/src/coreclr/pal/src/exception/seh-unwind.cpp:566
#4  0x0000ffffbdd8e480 in LazyMachState::unwindLazyState (baseState=<optimized out>, unwoundstate=0xffffffffc160, threadId=<optimized out>, funCallDepth=0, 
    hostCallPreference=<optimized out>) at /root/repos/runtime/src/coreclr/vm/arm64/stubs.cpp:340
#5  0x0000ffffbdbe1cac in HelperMethodFrame::InsureInit (this=0xffffffffdaf8, initialInit=false, unwindState=<optimized out>, 
    hostCallPreference=AllowHostCalls) at /root/repos/runtime/src/coreclr/vm/frames.cpp:1813
#6  0x0000ffffbdbe1c0c in HelperMethodFrame::GetFunction (this=0xffffffffdaf8) at /root/repos/runtime/src/coreclr/vm/frames.cpp:1732
#7  0x0000ffffbdd940fc in ExceptionTracker::InitializeCrawlFrameForExplicitFrame (pcfThisFrame=pcfThisFrame@entry=0xffffffffcd10, 
    pFrame=pFrame@entry=0xffffffffdaf8, pMD=pMD@entry=0xffff4432ff70) at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:1329
#8  0x0000ffffbdd93a58 in ExceptionTracker::ProcessOSExceptionNotification (this=this@entry=0xaaaaaab92a90, 
    pExceptionRecord=pExceptionRecord@entry=0xaaaaaac8dee0, pContextRecord=<optimized out>, pDispatcherContext=pDispatcherContext@entry=0xffffffffd288, 
    dwExceptionFlags=dwExceptionFlags@entry=0, sf=sf@entry=..., pThread=pThread@entry=0xaaaaaab94a90, STState=<optimized out>, 
    pICFSetAsLimitFrame=<optimized out>) at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:1863
#9  0x0000ffffbdd92cd0 in ProcessCLRException (pExceptionRecord=0xaaaaaac8dee0, MemoryStackFp=281474976701536, pContextRecord=0xaaaaaac8db50, 
    pDispatcherContext=pDispatcherContext@entry=0xffffffffd288) at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:1072
#10 0x0000ffffbdd96f6c in UnwindManagedExceptionPass1 (ex=..., frameContext=0xffffffffd6e0) at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:4624
#11 0x0000ffffbdd9725c in DispatchManagedException (ex=..., isHardwareException=<optimized out>)
    at /root/repos/runtime/src/coreclr/vm/exceptionhandling.cpp:4810
#12 0x0000ffffbdcfc188 in IL_Throw (obj=<optimized out>) at /root/repos/runtime/src/coreclr/vm/jithelpers.cpp:4024
#13 0x0000ffff445aa6a8 in ?? ()
#14 0x0000ffbf3680c858 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

@janvorli do you know what could be the cause of this regression?

> Was it built on RHEL9?

Yes, this was built on RHEL9.

> And can you please check what was the UNWIND_CONTEXT_IS_UCONTEXT_T detected by the configure.cmake?

```
./obj/coreclr/Linux.arm64.Release/pal/src/config.h:#define UNWIND_CONTEXT_IS_UCONTEXT_T 0
```

```
Performing C SOURCE FILE Test UNWIND_CONTEXT_IS_UCONTEXT_T failed with the following output:
Change Dir: /root/repos/runtime/artifacts/obj/coreclr/Linux.arm64.Release/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_716bb/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_716bb.dir/build.make CMakeFiles/cmTC_716bb.dir/build
gmake[1]: Entering directory '/root/repos/runtime/artifacts/obj/coreclr/Linux.arm64.Release/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_716bb.dir/src.c.o
/usr/bin/clang-13 -DUNWIND_CONTEXT_IS_UCONTEXT_T -D_FILE_OFFSET_BITS=64 -I/root/repos/runtime/src/native/external/libunwind/include -I/root/repos/runtime/artifacts/obj/external/libunwind/include -fPIE   -D_BSD_SOURCE -D_SVID_SOURCE -D_DEFAULT_SOURCE -D_POSIX_C_SOURCE=200809L -std=gnu11 -MD -MT CMakeFiles/cmTC_716bb.dir/src.c.o -MF CMakeFiles/cmTC_716bb.dir/src.c.o.d -o CMakeFiles/cmTC_716bb.dir/src.c.o -c /root/repos/runtime/artifacts/obj/coreclr/Linux.arm64.Release/CMakeFiles/CMakeTmp/src.c
/root/repos/runtime/artifacts/obj/coreclr/Linux.arm64.Release/CMakeFiles/CMakeTmp/src.c:8:26: error: assigning to 'unw_context_t' (aka 'unw_tdep_context_t') from incompatible type 'ucontext_t' (aka 'struct ucontext_t')
        libUnwindContext = uContext;
                         ^ ~~~~~~~~
1 error generated.
```

--- a/src/native/external/libunwind-version.txt
+++ b/src/native/external/libunwind-version.txt
@@ -7,6 +7,7 @@ Apply https://github.com/libunwind/libunwind/pull/330
 Apply https://github.com/libunwind/libunwind/pull/333
 Apply https://github.com/libunwind/libunwind/pull/342
 Apply https://github.com/libunwind/libunwind/pull/353
+Apply https://github.com/libunwind/libunwind/pull/358
 
 For LoongArch64:
 Apply https://github.com/libunwind/libunwind/pull/316 and https://github.com/libunwind/libunwind/pull/322

--- a/src/native/external/libunwind-version.txt
+++ b/src/native/external/libunwind-version.txt
@@ -3,8 +3,10 @@ https://github.com/libunwind/libunwind/commit/b3ca1b59a795a617877c01fe5d299ab7a0
 
 Reapply changes from https://github.com/dotnet/runtime/commit/1b5719c2e3dde393531eaeb5b5cde05dabeef5b8
 Apply https://github.com/libunwind/libunwind/pull/317
+Apply https://github.com/libunwind/libunwind/pull/330
 Apply https://github.com/libunwind/libunwind/pull/333
 Apply https://github.com/libunwind/libunwind/pull/342
+Apply https://github.com/libunwind/libunwind/pull/353
 
 For LoongArch64:
 Apply https://github.com/libunwind/libunwind/pull/316 and https://github.com/libunwind/libunwind/pull/322

--- a/src/native/external/libunwind/include/libunwind-aarch64.h
+++ b/src/native/external/libunwind/include/libunwind-aarch64.h
@@ -243,7 +243,7 @@ typedef ucontext_t unw_tdep_context_t;
      "stp x8, x9, [%[base], #64]\n" \
      "stp x10, x11, [%[base], #80]\n" \
      "stp x12, x13, [%[base], #96]\n" \
-     "stp x14, x13, [%[base], #112]\n" \
+     "stp x14, x15, [%[base], #112]\n" \
      "stp x16, x17, [%[base], #128]\n" \
      "stp x18, x19, [%[base], #144]\n" \
      "stp x20, x21, [%[base], #160]\n" \

--- a/src/native/external/libunwind/include/libunwind_i.h
+++ b/src/native/external/libunwind/include/libunwind_i.h
@@ -55,6 +55,8 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 #include <string.h>
 #include <unistd.h>
 #include <sys/mman.h>
+#include <errno.h>
+#include <stdio.h>
 
 #if defined(HAVE_ELF_H)
 # include <elf.h>
@@ -286,6 +288,13 @@ static ALWAYS_INLINE int
 print_error (const char *string)
 {
   return write (2, string, strlen (string));
+}
+
+HIDDEN extern long unw_page_size;
+
+static inline unw_word_t uwn_page_start(unw_word_t addr)
+{
+  return addr & ~(unw_page_size - 1);
 }
 
 #define mi_init         UNWI_ARCH_OBJ(mi_init)

--- a/src/native/external/libunwind/include/win/unistd.h
+++ b/src/native/external/libunwind/include/win/unistd.h
@@ -24,5 +24,8 @@ int          getpagesize(void);
 int          open(const char *, int, ...);
 ssize_t      read(int fd, void *buf, size_t count);
 ssize_t      write(int, const void *, size_t);
+long         sysconf(int name);
+
+#define _SC_PAGESIZE 11
 
 #endif // _MSC_VER

--- a/src/native/external/libunwind/src/aarch64/Ginit.c
+++ b/src/native/external/libunwind/src/aarch64/Ginit.c
@@ -84,8 +84,6 @@ get_dyn_info_list_addr (unw_addr_space_t as, unw_word_t *dyn_info_list_addr,
   return 0;
 }
 
-#define PAGE_SIZE 4096
-#define PAGE_START(a)   ((a) & ~(PAGE_SIZE-1))
 
 static int mem_validate_pipe[2] = {-1, -1};
 
@@ -197,11 +195,14 @@ tdep_init_mem_validate (void)
 
 #ifdef HAVE_MINCORE
   unsigned char present = 1;
-  unw_word_t addr = PAGE_START((unw_word_t)&present);
+  size_t len = unw_page_size;
+  unw_word_t addr = uwn_page_start((unw_word_t)&present);
   unsigned char mvec[1];
   int ret;
-  while ((ret = mincore ((void*)addr, PAGE_SIZE, (unsigned char *)mvec)) == -1 &&
-         errno == EAGAIN) {}
+  while ((ret = mincore((void *)addr, len, (unsigned char *)mvec)) == -1 &&
+         errno == EAGAIN)
+  {
+  }
   if (ret == 0)
     {
       Debug(1, "using mincore to validate memory\n");
@@ -295,12 +296,8 @@ validate_mem (unw_word_t addr)
 {
   size_t len;
 
-  if (PAGE_START(addr + sizeof (unw_word_t) - 1) == PAGE_START(addr))
-    len = PAGE_SIZE;
-  else
-    len = PAGE_SIZE * 2;
-
-  addr = PAGE_START(addr);
+  len = unw_page_size;
+  addr = uwn_page_start(addr);
 
   if (addr == 0)
     return -1;

--- a/src/native/external/libunwind/src/arm/Ginit.c
+++ b/src/native/external/libunwind/src/arm/Ginit.c
@@ -73,9 +73,6 @@ get_dyn_info_list_addr (unw_addr_space_t as, unw_word_t *dyn_info_list_addr,
   return 0;
 }
 
-#define PAGE_SIZE 4096
-#define PAGE_START(a)	((a) & ~(PAGE_SIZE-1))
-
 /* Cache of already validated addresses */
 #define NLGA 4
 static unw_word_t last_good_addr[NLGA];
@@ -85,14 +82,8 @@ static int
 validate_mem (unw_word_t addr)
 {
   int i, victim;
-  size_t len;
-
-  if (PAGE_START(addr + sizeof (unw_word_t) - 1) == PAGE_START(addr))
-    len = PAGE_SIZE;
-  else
-    len = PAGE_SIZE * 2;
-
-  addr = PAGE_START(addr);
+  size_t len = unw_page_size;
+  addr = uwn_page_start(addr);
 
   if (addr == 0)
     return -1;

--- a/src/native/external/libunwind/src/mi/init.c
+++ b/src/native/external/libunwind/src/mi/init.c
@@ -39,6 +39,30 @@ static const char rcsid[] UNUSED =
 long unwi_debug_level;
 
 #endif /* UNW_DEBUG */
+long unw_page_size;
+static void
+unw_init_page_size ()
+{
+  errno = 0;
+  long result = sysconf (_SC_PAGESIZE);
+  if (result == -1)
+    {
+      if (errno != 0)
+        {
+          print_error ("Failed to get _SC_PAGESIZE: ");
+          print_error (strerror(errno));
+          print_error ("\n");
+        }
+        else
+          print_error ("Failed to get _SC_PAGESIZE, errno was not set.\n");
+
+      unw_page_size = 4096;
+    }
+  else
+    {
+      unw_page_size = result;
+    }
+}
 
 HIDDEN void
 mi_init (void)
@@ -55,6 +79,6 @@ mi_init (void)
       setbuf (stderr, NULL);
     }
 #endif
-
-  assert (sizeof (struct cursor) <= sizeof (unw_cursor_t));
+  unw_init_page_size();
+  assert(sizeof(struct cursor) <= sizeof(unw_cursor_t));
 }

--- a/src/native/external/libunwind/src/riscv/Ginit.c
+++ b/src/native/external/libunwind/src/riscv/Ginit.c
@@ -97,9 +97,6 @@ get_dyn_info_list_addr (unw_addr_space_t as, unw_word_t *dyn_info_list_addr,
 
 // Memory validation routines are from aarch64
 
-#define PAGE_SIZE 4096
-#define PAGE_START(a)   ((a) & ~(PAGE_SIZE-1))
-
 static int mem_validate_pipe[2] = {-1, -1};
 
 #ifdef HAVE_PIPE2
@@ -210,11 +207,14 @@ tdep_init_mem_validate (void)
 
 #ifdef HAVE_MINCORE
   unsigned char present = 1;
-  unw_word_t addr = PAGE_START((unw_word_t)&present);
+  size_t len = unw_page_size;
+  unw_word_t addr = uwn_page_start((unw_word_t)&present);
   unsigned char mvec[1];
   int ret;
-  while ((ret = mincore ((void*)addr, PAGE_SIZE, (unsigned char *)mvec)) == -1 &&
-         errno == EAGAIN) {}
+  while ((ret = mincore((void *)addr, len, (unsigned char *)mvec)) == -1 &&
+         errno == EAGAIN)
+  {
+  }
   if (ret == 0)
     {
       Debug(1, "using mincore to validate memory\n");
@@ -306,14 +306,8 @@ cache_valid_mem(unw_word_t addr)
 static int
 validate_mem (unw_word_t addr)
 {
-  size_t len;
-
-  if (PAGE_START(addr + sizeof (unw_word_t) - 1) == PAGE_START(addr))
-    len = PAGE_SIZE;
-  else
-    len = PAGE_SIZE * 2;
-
-  addr = PAGE_START(addr);
+  size_t len = unw_page_size;
+  addr = uwn_page_start(addr);
 
   if (addr == 0)
     return -1;

--- a/src/native/external/libunwind/src/s390x/Ginit.c
+++ b/src/native/external/libunwind/src/s390x/Ginit.c
@@ -27,6 +27,7 @@ LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 
+#include "libunwind_i.h"
 #ifdef HAVE_CONFIG_H
 #include <config.h>
 #endif
@@ -92,9 +93,6 @@ get_dyn_info_list_addr (unw_addr_space_t as, unw_word_t *dyn_info_list_addr,
   *dyn_info_list_addr = _U_dyn_info_list_addr ();
   return 0;
 }
-
-#define PAGE_SIZE 4096
-#define PAGE_START(a)   ((a) & ~(PAGE_SIZE-1))
 
 static int mem_validate_pipe[2] = {-1, -1};
 
@@ -163,7 +161,7 @@ static int mincore_validate (void *addr, size_t len)
       return -1;
     }
 
-  for (i = 0; i < (len + PAGE_SIZE - 1) / PAGE_SIZE; i++)
+    for (i = 0; i < (len + unw_page_size - 1) / unw_page_size; i++)
     {
       if (!(mvec[i] & 1)) return -1;
     }
@@ -183,11 +181,14 @@ tdep_init_mem_validate (void)
 
 #ifdef HAVE_MINCORE
   unsigned char present = 1;
-  unw_word_t addr = PAGE_START((unw_word_t)&present);
+  size_t len = unw_page_size;
+  unw_word_t addr = uwn_page_start((unw_word_t)&present);
   unsigned char mvec[1];
   int ret;
-  while ((ret = mincore ((void*)addr, PAGE_SIZE, mvec)) == -1 &&
-         errno == EAGAIN) {}
+  while ((ret = mincore((void *)addr, len, mvec)) == -1 &&
+         errno == EAGAIN)
+  {
+  }
   if (ret == 0 && (mvec[0] & 1))
     {
       Debug(1, "using mincore to validate memory\n");
@@ -210,14 +211,8 @@ static int
 validate_mem (unw_word_t addr)
 {
   int i, victim;
-  size_t len;
-
-  if (PAGE_START(addr + sizeof (unw_word_t) - 1) == PAGE_START(addr))
-    len = PAGE_SIZE;
-  else
-    len = PAGE_SIZE * 2;
-
-  addr = PAGE_START(addr);
+  size_t len = unw_page_size;
+  addr = uwn_page_start(addr);
 
   if (addr == 0)
     return -1;

--- a/src/native/external/libunwind/src/win/pal-single-threaded.c
+++ b/src/native/external/libunwind/src/win/pal-single-threaded.c
@@ -112,6 +112,13 @@ ssize_t read(int fd, void *buf, size_t count)
     return -1;
 }
 
+ssize_t write(int fd, const void *buf, size_t nbyte)
+{
+    // For dump debugging we shouldn't need to open files
+    // Especially since we didn't implement open()
+    return -1;
+}
+
 int close(int fd)
 {
     // For dump debugging we shouldn't need to open files

--- a/src/native/external/libunwind/src/win/pal-single-threaded.c
+++ b/src/native/external/libunwind/src/win/pal-single-threaded.c
@@ -15,6 +15,7 @@
 #include <sys/mman.h>
 #include <sys/stat.h>
 #include <unistd.h>
+#include <errno.h>
 #include "libunwind_i.h"
 #include "compiler.h"
 
@@ -23,6 +24,17 @@ int getpagesize(void)
     // 4096 is truth for most targets
     // Unlikely to matter in dump debugging
     return 4096;
+}
+
+long sysconf(int name)
+{
+    if (name == _SC_PAGESIZE)
+    {
+        return getpagesize();
+    }
+
+    errno = EINVAL;
+    return -1;
 }
 
 void* mmap(void *addr, size_t length, int prot, int flags, int fd, size_t offset)

--- a/src/native/external/libunwind/src/x86/Ginit.c
+++ b/src/native/external/libunwind/src/x86/Ginit.c
@@ -74,9 +74,6 @@ get_dyn_info_list_addr (unw_addr_space_t as, unw_word_t *dyn_info_list_addr,
   return 0;
 }
 
-#define PAGE_SIZE 4096
-#define PAGE_START(a)   ((a) & ~(PAGE_SIZE-1))
-
 /* Cache of already validated addresses */
 #define NLGA 4
 static unw_word_t last_good_addr[NLGA];
@@ -89,14 +86,8 @@ validate_mem (unw_word_t addr)
 #ifdef HAVE_MINCORE
   unsigned char mvec[2]; /* Unaligned access may cross page boundary */
 #endif
-  size_t len;
-
-  if (PAGE_START(addr + sizeof (unw_word_t) - 1) == PAGE_START(addr))
-    len = PAGE_SIZE;
-  else
-    len = PAGE_SIZE * 2;
-
-  addr = PAGE_START(addr);
+  size_t len = unw_page_size;
+  addr = uwn_page_start(addr);
 
   if (addr == 0)
     return -1;

--- a/src/native/external/libunwind/src/x86_64/Ginit.c
+++ b/src/native/external/libunwind/src/x86_64/Ginit.c
@@ -73,9 +73,6 @@ get_dyn_info_list_addr (unw_addr_space_t as, unw_word_t *dyn_info_list_addr,
   return 0;
 }
 
-#define PAGE_SIZE 4096
-#define PAGE_START(a)   ((a) & ~(PAGE_SIZE-1))
-
 static int mem_validate_pipe[2] = {-1, -1};
 
 #ifdef HAVE_PIPE2
@@ -191,10 +188,11 @@ tdep_init_mem_validate (void)
 
 #ifdef HAVE_MINCORE
   unsigned char present = 1;
-  unw_word_t addr = PAGE_START((unw_word_t)&present);
+  size_t len = unw_page_size;
+  unw_word_t addr = uwn_page_start((unw_word_t)&present);
   unsigned char mvec[1];
   int ret;
-  while ((ret = mincore ((void*)addr, PAGE_SIZE, (unsigned char *)mvec)) == -1 &&
+  while ((ret = mincore ((void*)addr, len, (unsigned char *)mvec)) == -1 &&
          errno == EAGAIN) {}
   if (ret == 0)
     {
@@ -287,14 +285,8 @@ cache_valid_mem(unw_word_t addr)
 static int
 validate_mem (unw_word_t addr)
 {
-  size_t len;
-
-  if (PAGE_START(addr + sizeof (unw_word_t) - 1) == PAGE_START(addr))
-    len = PAGE_SIZE;
-  else
-    len = PAGE_SIZE * 2;
-
-  addr = PAGE_START(addr);
+  size_t len = unw_page_size;
+  addr = uwn_page_start(addr);
 
   if (addr == 0)
     return -1;

